dbms/src/IO/ReadHelpers.h

```diff
@@ -911,4 +912,10 @@ void skipToNextLineOrEOF(ReadBuffer & buf);
 /// Skip to next character after next unescaped \n. If no \n in stream, skip to end. Does not throw on invalid escape sequences.
 void skipToUnescapedNextLineOrEOF(ReadBuffer & buf);
 
+/** Return buffer eof() result.
+  * If there is no pending data in buffer or it was explicitly asked
+  *  save current state to memory.
+  */
+bool eofWithSavingBufferState(ReadBuffer & buf, DB::Memory<> & memory, size_t & used_size, char * & begin_pos, bool save_buffer_state = false);
```

Comment:
My mind just goes blank when I see this function and how it's used. At the very least, it should have a sane name, and no bool flag that modifies behavior.